connector.go

```diff
@@ -0,0 +1,133 @@
+// Go MySQL Driver - A MySQL-Driver for Go's database/sql package
+//
+// Copyright 2018 The Go-MySQL-Driver Authors. All rights reserved.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package mysql
+
+import (
+	"context"
+	"database/sql/driver"
+	"net"
+)
+
+// Connect implements driver.Connector interface.
+// Connect returns a connection to the database.
+func (c *Config) Connect(ctx context.Context) (driver.Conn, error) {
+	var err error
+
+	// New mysqlConn
+	mc := &mysqlConn{
+		maxAllowedPacket: maxPacketSize,
+		maxWriteSize:     maxPacketSize - 1,
+		closech:          make(chan struct{}),
+		cfg:              c,
+	}
+	mc.parseTime = mc.cfg.ParseTime
+
+	// Connect to Server
+	// TODO: needs RegisterDialContext
+	dialsLock.RLock()
+	dial, ok := dials[mc.cfg.Net]
+	dialsLock.RUnlock()
+	if ok {
+		mc.netConn, err = dial(mc.cfg.Addr)
+	} else {
+		nd := net.Dialer{Timeout: mc.cfg.Timeout}
+		mc.netConn, err = nd.DialContext(ctx, mc.cfg.Net, mc.cfg.Addr)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Enable TCP Keepalives on TCP connections
+	if tc, ok := mc.netConn.(*net.TCPConn); ok {
+		if err := tc.SetKeepAlive(true); err != nil {
+			// Don't send COM_QUIT before handshake.
+			mc.netConn.Close()
+			mc.netConn = nil
+			return nil, err
+		}
+	}
+
+	// Call startWatcher for context support (From Go 1.8)
+	if s, ok := interface{}(mc).(watcher); ok {
+		s.startWatcher()
+		if err := s.watchCancel(ctx); err != nil {
```

Comment:
This looks new during `Connect`. What does it do?

Reply:
It cancels opening new connection, if the context is canceled.
See https://github.com/go-sql-driver/mysql/pull/608 please.

Other methods which supports the context have this lines. (e.g. https://github.com/go-sql-driver/mysql/blob/2307b45d3a41dcc64a2bf46db80f61fd182e0186/connection_go18.go#L84-L87)
I did not need it because `MySQLDriver.Open` did not correspond to the context so far, but since `Config.Connect` become to correspond to the context, I added this